[ROCm] ROCm support for MLIR generated kernels

deven-amd opened the pull request:
This PR contains prototype code to add ROCm support for MLIR generated kernels.

The intent behind this PR is to share what we have done and invite feedback on whether or not we are going down the right path, and what changes we need to make.

Reviewing the PR on a per commit basis will work best

Note this is prototype code, which is half-baked in that it is hardcoded to generate GPU binaries for the "gfx900" architecture, and does not yet have the ability to support gpu binaries for more than one gpu_arch. However it is fully functional, and the unit test `//tensorflow/core/kernels/mlir_generated:gpu_tanh_test` will successfully run and pass with it on the ROCm platform

--------------------------

/cc @whchung @chsigg @cheshire @nvining-work

deven-amd committed:
Adding a LowerKernelBodiesToROCDL function which lowers MLIR GPU Modules to ROCDL Dialect

deven-amd committed:
Rename GenerateCubinForTfCode to GenerateGpuBinaryForTfCode and add ROCm support for it

deven-amd committed:
Adding ROCM support for the code within "tensorflow/core/kernels/mlir_generated/"

deven-amd committed:
Update configure.py to provide a way to enable MLIR generated gpu kernels on the ROCm platform.

MLIR generated gpu kernels will continue to be disabled by default.

This commit provides a way to enable them, if conifugre.py is run with the env `TF_ENABLE_MLIR_GENERATED_GPU_KERNELS=1`

deven-amd committed:
renaming cubin_creator* to gpu_binary_creator*. This is to reflect the fact that the utility can now be used to generate GPU binaries for both AMD And NVIDIA Gpus

akuegel commented on configure.py:
```diff
@@ -1442,6 +1442,9 @@ def main():
     write_action_env_to_bazelrc('ROCM_PATH', environ_cp.get('ROCM_PATH'))
     write_action_env_to_bazelrc('ROCM_ROOT', environ_cp.get('ROCM_PATH'))
 
+  if environ_cp.get('TF_NEED_ROCM') == '1':
```
I guess this was just for easier testing? Although we also want to flip the flag soon for cuda devices, there are still some blockers for that, for example it seems we still use the old kernels in some tensorflow codepaths even when enabling this flag, and then there can be small numeric differences in the results. But maybe this is not such a big deal, because these differences would also show up when running on different backends. We are still discussing this internally whether it is really a blocker, and what we can do about it.

deven-amd replied:
>I guess this was just for easier testing?

yes. In the long run, the expectation would be that this flag would have the same default value in ROCm and CUDA.

akuegel commented on tensorflow/compiler/mlir/tools/kernel_gen/tf_to_gpu_binary.cc:
```diff
@@ -64,8 +64,13 @@ int main(int argc, char** argv) {
     return 1;
   }
 
+#if TENSORFLOW_USE_ROCM
```
Since both methods share the same first few lowering passes and have the same signature, it might make sense to have a common function to be called and do the specialization for Hsaco code later?

deven-amd replied:
Yes. we probably should have a single `GenerateGpuBinaryForTfCode` routine / call here, and the implementation of that routine, can have some ROCm / CUDA specific aspects to it. I will make that change in the next iteration. For this iteration, one of the goals was to leave the CUDA path as untouched as possible, to help make the rebasing easier.

akuegel commented on tensorflow/compiler/mlir/tools/kernel_gen/gpu_binary_creator.cc:
```diff
@@ -309,3 +313,49 @@ StatusOr<std::vector<uint8_t>> tensorflow::kernel_gen::GenerateCubinForTfCode(
+    TF_RETURN_IF_ERROR(xla::mlir_gpu::LowerLHLOToGPU(module.get(), options));
+  }
+
+  TF_RETURN_IF_ERROR(xla::mlir_gpu::LowerKernelBodiesToROCDL(module.get()));
```
If I see it correctly, only after here we do something different for cuda and rocm.

akuegel commented on tensorflow/core/kernels/mlir_generated/build_defs.bzl:
```diff
@@ -47,10 +52,40 @@ def _gen_kernel_cubin_impl(ctx):
         cubins.append(cubin)
     return [CubinInfo(cubins = cubins)]
 
+def _gen_kernel_cubin_impl_rocm(ctx):
```
I guess this could be shared between both cuda and rocm, if we move the line
```
arch = arch.replace("compute_", "sm_")
```
to the cuda-specific calling function.

deven-amd replied:
correct

akuegel commented on tensorflow/core/kernels/mlir_generated/build_defs.bzl:
```diff
@@ -101,29 +137,41 @@ def _gen_kernel_image_hdr_impl(ctx):
+    ctx.actions.run_shell(
+        outputs = [ctx.outputs.out],
+        inputs = hsacos,
+        command = "echo 'const std::string %s = \"/root/tensorflow/%s\";' > %s" % (ctx.attr.symbol, hsacos[0].path, ctx.outputs.out.path),
```
I guess this is the prototype aspect of this PR? :)
In the end we probably need to find a different way, because the assumption that the code root is always /root/tensorflow is probably not valid?

deven-amd replied:
yes.  The prototype aspect primarily involves a couple of things

1. We need to implement the equivalent of `cuda_gpu_architectures` for ROCm. That will involve passing that information in via rocm_configure.bzl / local_config_rocm. I expect this to be present in the next iteration as well

2. We need to figure out how to implement the equivalent of `fatbinary` and `bin2c` on the ROCm platform. The prototype code
 * supports generating HSACOs (ROCm GPU Binaries) for a single architecture only (gfx900) 
  *need something like `fatbinary` to support more than arch 
 * needs to pass the harcoded file name of the HSACO file, because of the lack of a `bin2c` type utility

We are still looking into how to make this part work on ROCm. Any suggestions / feedback is welcome.

akuegel reviewed (commented):
Thanks for working on this :)
In general I think this is going in the right direction, and thanks also for doing a bit of cleanup with indentation / naming / ordering.
FYI, I am currently working on reducing the amount of boilerplate code for registering a MLIR generated kernel, and using that to also add the "abs" kernel. This may then require a few adjustments in cwise_op_gpu_tanh.cu.cc.

deven-amd reviewed (commented):
hi Adrian,

Thank you for your prompt feedback. will incorporate it in the next iteration of the prototype. 
Please cc me when you push out the changes for `abs`, and I will make the corresponding ROCm updates in my prototype as well.

thanks

akuegel commented:
> hi Adrian,
> 
> Thank you for your prompt feedback. will incorporate it in the next iteration of the prototype.
> Please cc me when you push out the changes for `abs`, and I will make the corresponding ROCm updates in my prototype as well.
> 
> thanks

This change has now landed:
https://github.com/tensorflow/tensorflow/commit/a6cd18a1334be6b54784f1d65fe5d435d31d2bb8

deven-amd commented:
> > hi Adrian,
> > Thank you for your prompt feedback. will incorporate it in the next iteration of the prototype.
> > Please cc me when you push out the changes for `abs`, and I will make the corresponding ROCm updates in my prototype as well.
> > thanks
> 
> This change has now landed:
> [a6cd18a](https://github.com/tensorflow/tensorflow/commit/a6cd18a1334be6b54784f1d65fe5d435d31d2bb8)

@akuegel 
Thank you for the heads up. I will update this prototype to incorporate the changes in the above commit (sometime next week, when our fork syncs up with the upstream repo and picks up the above commit in the process)

@gbaned 
The merge-conflicts will be gone once I update the prototype as noted above and push out that commit
Keep in mind that (atleast for now) this PR represents prototype code that is not intended to be merged. Once we get to a stage where the changes in this PR are clean enough, I will remove the `[prototype]` prefix from the PR title. 

thanks

deven-amd committed:
Updating how amdgpu_targets are determined in rocm_configure.bzl

Prior to this commit, the AMD GPU targets (i.e. the `amdgpu_targets`), for which HSACO objects are created in the TF build, were determined as follows.
* No `--amdgpu-target=` option would be explicitly added to the `hipcc` command line (via `rocm_copts`)
* `hipcc` would, upon not seeing any `--amdgpu-target=` option specified, invoke the `$ROCM_PATH/bin/rocm_agent_enumerator` tool to determine the list of `amdgpu_targets`

This commit moves the determination of `amdgpu_targets` to be in `rocm_configure.bzl`. Instead of in `hipcc`, the `$ROCM_PATH/bin/rocm_agent_enumerator` tool will instead be invoked within `rocm_configure.bzl` to determine the list `amdgpu_targets`. For each `target` in the `amdgpu_targets` list, a `--amdgpu-target=<target>` option will be added to the `hipcc` command line (via `rocm_copts()`).

This commit also
* allows overriding the way `amdgpu_targets` are determined, by setting the env var `TF_ROCM_AMDGPU_TARGETS` instead.
* creates `rocm_gpu_archictectures` routine in `@local_config_rocm/build_defs.bzl`, which returns the `amgpu_targets` list.
  * This will come in the handy when determining the `amdgpu_targets` to build for, when compiling MLIR generated kernels, using the XLA backend (in the non XLA path)

deven-amd committed:
Copying the clang-offload-bundler binary to local_config_rocm repo

deven-amd committed:
Add the bin2c.py utility and copy it to the local_config_rocm repo

akuegel commented:
Sorry, I was on vacation the last two weeks, and returned yesterday. Still catching up on emails.
It seems in the meantime there were some changes which have again lead to merge conflicts, can you please resolve these?
Btw, by now the MLIR generated kernels for abs and tanh are enabled by default for the cuda code path, and this seems to have stuck. So it looks like this will become part of the next tensorflow release.

deven-amd commented:
@akuegel 

I have rebased the PR to resolve the merge conflicts. Please re-review. 

You mentioned that these MLIR generated kernels (for tanh and abs) will become default (on the cuda path) in the next release. Can you shed some light on why we are going down this route (of using MLIR generated kernels on GPU, in the non-XLA path)? 

Is the eventual goal here to replace all/most Eigen generated kernels with MLIR generated ones?

akuegel commented:
> You mentioned that these MLIR generated kernels (for tanh and abs) will become default (on the cuda path) in the next release. Can you shed some light on why we are going down this route (of using MLIR generated kernels on GPU, in the non-XLA path)?
> 
> Is the eventual goal here to replace all/most Eigen generated kernels with MLIR generated ones?

I don't know how much I can share, but there will be a public tech talk about this which was announced as part of the MLIR newsletter:
https://llvm.discourse.group/t/mlir-news-13th-edition-8-7-2020/1478
(section Tensorflow - Kernel Generator)

akuegel commented on tensorflow/compiler/xla/service/hlo_module_config.cc:
```diff
@@ -51,12 +51,14 @@ string HloModuleConfig::compilation_cache_key() const {
-  for (const ShapeLayout& param_layout :
-       entry_computation_layout_->parameter_layouts()) {
-    params.push_back(param_layout.shape().DebugString());
+  if (entry_computation_layout_.has_value()) {
```
I might be wrong, but this change seems unrelated. If yes, please create a separate pull request for it.

joker-eph commented:
> Can you shed some light on why we are going down this route (of using MLIR generated kernels on GPU, in the non-XLA path)?

There are a few reasons, but: we could express TF ops as composition and just reuse the code generator when building TF to have the kernel generated AOT for these compiled op. Ultimately having the JIT using the same code path as the AOT codegen for most computations reduces the risk of divergence and make it more transparent to switch from one to another.

akuegel commented:
It seems there is yet another merge conflict in cubin_creator.cc, can you please resolve this?

deven-amd commented:
@akuegel 

I have rebased the PR to resolve the merge-conflict.
The unfortunate side effect of that is that the two associated unit-tests (abs and tanh) are no longer passing after the rebase. They both crash during the phase that genrates GPU kernel code objects from the input MLIR. The crash is in the MLIR back end, which has changed with the rebase.

Was wondering whether this crash is ROCm specific or is it something you are seeing on the CUDA side too?

thanks

error message and stack trace at the point of failure
```
ERROR: /root/tensorflow/tensorflow/core/kernels/mlir_generated/BUILD:123:1: compile tensorflow/core/kernels/mlir_generated/abs_f16_kernel_cubin.gfx803.hsaco failed (Segmentation fault): tf_to_gpu_binary failed: error executing command 
  (cd /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/execroot/org_tensorflow && \
  exec env - \
  bazel-out/host/bin/tensorflow/compiler/mlir/tools/kernel_gen/tf_to_gpu_binary '--same_shape=0,1' '--unroll_factors=4' '--tile_sizes=256' '--arch=803' '--input=bazel-out/k8-opt/bin/tensorflow/core/kernels/mlir_generated/abs_f16.mlir' '--output=bazel-out/k8-opt/bin/tensorflow/core/kernels/mlir_generated/abs_f16_kernel_cubin.gfx803.hsaco')
Execution platform: @local_execution_config_platform//:platform
PLEASE submit a bug report to  and include the crash backtrace.
```



```
(gdb) where
#0  0x0000555558d32a87 in mlir::detail::OperandStorage::OperandStorage(mlir::Operation*, mlir::ValueRange) ()
#1  0x0000555558d3585a in mlir::Operation::create(mlir::Location, mlir::OperationName, llvm::ArrayRef<mlir::Type>, llvm::ArrayRef<mlir::Value>, mlir::MutableDictionaryAttr, llvm::ArrayRef<mlir::Block*>, unsigned int) ()
#2  0x0000555558d35964 in mlir::Operation::create(mlir::Location, mlir::OperationName, llvm::ArrayRef<mlir::Type>, llvm::ArrayRef<mlir::Value>, mlir::MutableDictionaryAttr, llvm::ArrayRef<mlir::Block*>, mlir::RegionRange) ()
#3  0x0000555558d35ba7 in mlir::Operation::create(mlir::OperationState const&) ()
#4  0x0000555558d8b684 in mlir::OpBuilder::createOperation(mlir::OperationState const&) ()
#5  0x0000555557facacb in mlir::GPUIndexIntrinsicOpLowering<mlir::gpu::BlockIdOp, mlir::ROCDL::BlockIdXOp, mlir::ROCDL::BlockIdYOp, mlir::ROCDL::BlockIdZOp>::matchAndRewrite(mlir::Operation*, llvm::ArrayRef<mlir::Value>, mlir::ConversionPatternRewriter&) const ()
#6  0x0000555558af4f13 in mlir::ConversionPattern::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const ()
#7  0x0000555558d284cb in mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::RewritePattern const&, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::RewritePattern const&)>, llvm::function_ref<void (mlir::RewritePattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::RewritePattern const&)>) ()
#8  0x0000555558d28688 in mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::RewritePattern const&)>, llvm::function_ref<void (mlir::RewritePattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::RewritePattern const&)>) ()
#9  0x0000555558afe3a9 in (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, mlir::ConversionPatternRewriter&) ()
#10 0x0000555558afef36 in (anonymous namespace)::OperationConverter::convertOperations(llvm::ArrayRef<mlir::Operation*>) ()
#11 0x0000555558b0035c in mlir::applyFullConversion(llvm::ArrayRef<mlir::Operation*>, mlir::ConversionTarget&, mlir::OwningRewritePatternList const&) ()
#12 0x0000555558b00470 in mlir::applyFullConversion(mlir::Operation*, mlir::ConversionTarget&, mlir::OwningRewritePatternList const&) ()
#13 0x0000555557d1d330 in xla::mlir_gpu::(anonymous namespace)::LowerToROCDLPass::runOnOperation() ()
#14 0x0000555558c51d08 in mlir::Pass::run(mlir::Operation*, mlir::AnalysisManager) ()
#15 0x0000555558c51dda in mlir::OpPassManager::run(mlir::Operation*, mlir::AnalysisManager) ()
#16 0x0000555558c51f2e in mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl()::{lambda(llvm::MutableArrayRef<mlir::OpPassManager>)#1}::operator()(llvm::MutableArrayRef<mlir::OpPassManager>) const ()
#17 0x0000555558c5731e in mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl() ()
#18 0x0000555558c51d08 in mlir::Pass::run(mlir::Operation*, mlir::AnalysisManager) ()
#19 0x0000555558c51dda in mlir::OpPassManager::run(mlir::Operation*, mlir::AnalysisManager) ()
#20 0x0000555558c595f2 in mlir::PassManager::run(mlir::ModuleOp) ()
#21 0x0000555557d1e0fe in xla::mlir_gpu::LowerKernelBodiesToROCDL(mlir::ModuleOp) ()
#22 0x0000555555a7e736 in tensorflow::kernel_gen::GenerateGpuBinaryForTfCode(llvm::StringRef, std::pair<int, int>, llvm::ArrayRef<unsigned int>, llvm::ArrayRef<unsigned int>, llvm::ArrayRef<unsigned int>) ()
#23 0x00005555559f6d7d in main ()
(gdb) quit

```

deven-amd commented:
@akuegel 

found the cause of crash....was failing to account (on the ROCDL side) the following change that got picked up when I rebased the PR on the tip.

https://github.com/tensorflow/tensorflow/commit/bcfb691a79a9e9c742ee9e4a7bf5af0a9edb1746#diff-d0fe1cfb091a0f7a5f03b710558059d7

I have made the correponding changes on the ROCDL side, and the two unit tests are back to passing.

akuegel reviewed (approved):
Thanks :)

gbaned commented:
@deven-amd  Can you please check build failures. Thanks!

akuegel commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -621,6 +634,7 @@ def _create_local_rocm_repository(repository_ctx):
                 repository_ctx,
                 rocm_config.amdgpu_targets,
             ),
+            "%{rocm_gpu_architectures}": str(rocm_config.amdgpu_targets),
```
You need a corresponding entry in create_dummy_repository. Otherwise it will not work in cuda builds.
I had the reverse problem when working on cuda support to prevent it failing on rocm.

deven-amd replied:
you are right!

have added the corresponding entry in the create_dummy_repository and pushed it out. thanks.

deven-amd commented:
@akuegel 

pushed out another rebase to resolve the `buildifier` errors in `Ubuntu Sanity`....please re-approve. thanks

akuegel commented:
For some reason, it seems this fails on the ROCM Kokoro. The problem is that it fails to find the file bin2c.py on the path returned by repository_ctx.path(Label("//third_party/gpus/rocm:bin2c.py"))
I tried the whole day all kind of stuff whether I can make it work, but failed. Maybe you have an idea? My guess is that this is somehow related to remote execution, but I don't really know. And the only way I could try it out was to repeatedly start another Kokoro run (and then see it fail).

deven-amd commented:
> Maybe you have an idea? My guess is that this is somehow related to remote execution, but I don't really know.

I do not quite understand how the remote exectution part works either :(
@chsigg would you happen to know? thanks

akuegel commented:
> > Maybe you have an idea? My guess is that this is somehow related to remote execution, but I don't really know.
> 
> I do not quite understand how the remote exectution part works either :(
> @chsigg would you happen to know? thanks

I got some help internally. Essentially remote execution doesn't support file upload yet. I am now trying whether we can just use hexdump in the genrule, so that we don't need the bin2c tool.
